implementation tests

Konrad1w opened the pull request:
I attempted to run CategoryManagementControllerTest and encountered an error which I previously mentioned to you.

Konrad1w committed:
implementation tests

Commit:
disable sec for tests example

dawba commented:
@Konrad1w could you provide some example of the error you've encountered?
I have checked out the build logs and its 68 PASS / 17 FAIL, which is a nice result. 

I will take a look at the tests you've implemented and see if there is something more to add. 
For now, the coverage looks great given only a few entities to test.

dimmonn commented on src/test/java/org/marketplace/controllers/UserManagementControllerTest.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Test
+    void getAllUsers() {
```
this and below are yet to be done

dimmonn commented on src/test/java/org/marketplace/services/AdvertisementImageManagementServiceTest.java:
```diff
@@ -0,0 +1,152 @@
+    @Test
+    void testAddImage_ShouldAddImage() {
+        // given
+        given(advertisementImageManagementRepository.findById(advertisementImage.getId())).willReturn(Optional.empty());
```
i would add real workflow here, there is no need to mock it up, we can just add stuff to h2 before each test
normally we want to use mocks as an exceptional situation

dawba replied:
Looking at the advertisement image this is a part of the project we failed at and have not fully implemented it, which is a bit of a pity.

Will be included as something we can learn from, in conclusions of this project.

dimmonn reviewed (commented):
we can keep current version just to save time 👍
in a real life project i would use mocks as an exceptional situation when things are really out of our control and therefore we want to mock up, in 99% of cases we can test based on abstractions/interfaces etc

Konrad1w committed:
add Controller tests for User and Category

Konrad1w commented:
I added tests for the user and category controllers. Wen trying to add tests for the advertisement controller, I encountered an error related to Java 8 LocalDateTime. According to the compiler's suggestion, I added com.fasterxml.jackson.datatype, but it didn't help, so I reverted that solution. Can somebody look at it?

Konrad1w committed:
make working ad controller test

Konrad1w commented:
> I added tests for the user and category controllers. Wen trying to add tests for the advertisement controller, I encountered an error related to Java 8 LocalDateTime. According to the compiler's suggestion, I added com.fasterxml.jackson.datatype, but it didn't help, so I reverted that solution. Can somebody look at it?

Ok I solved this problem. It was connected with version Jackson, which I added. Here is solution 
![image](https://github.com/dawba/market-place/assets/130360173/79711999-b67b-431b-9c7c-046b0278e80f)

dawba commented on src/main/java/org/marketplace/services/AdvertisementManagementService.java:
```diff
@@ -28,11 +28,16 @@ public AdvertisementManagementService(AdvertisementManagementRepository advertis
-            throw new EntityExistsException(String.format("Advertisement with id: %d already exsits!", advertisement.getId()));
-        } catch (EntityNotFoundException e) {
+        Long advertisementId = advertisement.getId();
+        if (advertisementId == null) {
```
We dont have to include an "if-else" block here.
The if clause is sufficient - just do a try-catch blok after the if statement, as when IF is not satisfied, we will execute the rest of the code.

dawba commented on src/main/java/org/marketplace/models/Advertisement.java:
```diff
@@ -42,6 +44,8 @@ public class Advertisement {
     @Column(name = "observer")
     private List<String> observers;
+    @JsonIgnore
+    private Long owner;
```
I am not seeing any usage of it, is it really needed? 
We have the owner based on the `Advertisement.user.id` field

Konrad1w replied:
I had to add it, because I got error running 'TestUtil.extractAdvertisementIdFromMvcResult' method. Error was connected with mapping Advertisement.

dawba commented on src/test/java/org/marketplace/controllers/AdvertisementManagementControllerTest.java:
```diff
@@ -18,26 +33,116 @@
+    private Long categoryId;
+
+    @Test
+    public void createAdvertisement() throws Exception {
```
Given the constraints on advertisement entity, I would test the border values for title/description fields, or passing the userId/categoryId as nullish.

dawba commented on src/test/java/org/marketplace/controllers/AdvertisementManagementControllerTest.java:
```diff
@@ -18,26 +33,116 @@
-        String payload = "{\"id\":1,\"title\":\"SampleAdvertisement\",\"description\":\"Thisisasampleadvertisement\",\"category\":{\"id\":1,\"name\":\"SampleCategory\"},\"user\":{\"id\":1,\"username\":\"SampleUser\"},\"price\":100.0,\"location\":\"SampleLocation\"}";
-
-        mockMvc.perform(post("/api/advertisement/add")
+        //String payload = "{\"id\":1,\"title\":\"SampleAdvertisement\",\"description\":\"Thisisasampleadvertisement\",\"category\":{\"id\":1,\"name\":\"SampleCategory\"},\"user\":{\"id\":1,\"username\":\"SampleUser\"},\"price\":100.0,\"location\":\"SampleLocation\"}";
```
If not used, I would just delete it I guess.

dawba commented on src/test/java/org/marketplace/controllers/UserManagementControllerTest.java:
```diff
@@ -0,0 +1,209 @@
+@AutoConfigureMockMvc
+@ActiveProfiles(value = "test")
+@Transactional
+class UserManagementControllerTest {
```
also:
maybe test the flow of

`register -> confirm -> login` to check the isVerified field and user access.

dawba replied:
Also, the constraints on phone/login length would be nice to be tested

dawba commented on src/test/java/org/marketplace/services/CategoryManagementServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    @Test
+    public void testAddCategoryWithNullName() {
```
given that we test the constraints here - why not test the border values of the category?
so drop in the 
names of length 1, 2, 20 and 21 + null
so we get it done in one test.

dawba commented on src/test/java/org/marketplace/services/CategoryManagementServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    @Test
+    public void testAddCategoryWithShortName() {
```
I know that we test it here, but maybe just do it in one single test, and then we can test other methods from the service?

dawba commented on src/test/java/org/marketplace/util/TestUtil.java:
```diff
@@ -0,0 +1,65 @@
+import org.marketplace.requests.Response;
+import org.springframework.test.web.servlet.MvcResult;
+
+public class TestUtil {
```
Couldnt we make it more reusable with passed generic type?
and do a switchcase here if you need a different logic for advertisement id.

Otherwise we duplicate our code a little bit too much.

so I would see something like

```java
    public static <T> Long extractIdFromMvcResult(MvcResult result, TypeReference<Response<T>> typeReference) {
        try {
            String responseBody = result.getResponse().getContentAsString();
            Response<T> response = mapper.readValue(responseBody, typeReference);
            return response.getData().getId();
        } catch (Exception e) {
            throw new RuntimeException("Failed to extract ID from MvcResult", e);
        }
    }

    public static Long extractAdvertisementIdFromMvcResult(MvcResult result) {
        return extractIdFromMvcResult(result, new TypeReference<Response<Advertisement>>() {});
    }

    public static Long extractUserIdFromMvcResult(MvcResult result) {
        return extractIdFromMvcResult(result, new TypeReference<Response<User>>() {});
    }

    public static Long extractCategoryIdFromMvcResult(MvcResult result) {
        return extractIdFromMvcResult(result, new TypeReference<Response<Category>>() {});
    }

    public static Long extractAdvertisementImageIdFromMvcResult(MvcResult result) {
        return extractIdFromMvcResult(result, new TypeReference<Response<AdvertisementImage>>() {});
    }
    ```
    
    or maybe it is possible to just make it generic and just call `extractIdFromMvcResult`

Konrad1w replied:
`return response.getData().getId();` this makes compile error, because we cannot use method getId on genericType. I   don't know if it is worth to make Interface with method getId(), which will be implemented by our Entities like User/Controller ... In this case it will be possible to make only one method to extractId

dawba reviewed (changes requested):
Great job! 

I like how the refreshed version looks like and feel like we test most of what is needed.

Please take a look at requested changes - and when finished I will get to creating a testing raport for our app.

Konrad1w committed:
reformat code

Konrad1w committed:
apply some suggested changes
